### PR TITLE
[ENH] add time series agglomerative clusterer

### LINKED ISSUE #3433: [ENH] Implement Agglomerative clustering algorithm compatible with aeon distances
### Describe the feature or idea you want to propose

see #1241 

this is assigned to SotonSweetXX

nice to have a time series specific algomorative and ultimately also top down hierarchical clusterer. 

so design choices:
1. ignore ward option for now
2. start with a precomputed distance only, then expand for others
3. Use either scipy.cluster.hierarchy.linkage or scikit-learn with precomputed, do a speed test to decide which
Decide on whether to use predict or not, some research on that required

should extend BaseClusterer

### Describe your proposed solution

something along the lines of

```python
class TimeSeriesAgglomerative(BaseClusterer):
    """Agglomerative (bottom-up) hierarchical clustering for time series."
    rest here
    """

    _tags = {
        "capability:multivariate": True,
        "algorithm_type": "distance",
    }

    def init(
        self,
        n_clusters: int | None = 8,
        distance_threshold: float | None = None,
        linkage: str = "average",
        distance: str | Callable = "msm",
        distance_params: dict | None = None,
    ):
        self.n_clusters = n_clusters
        self.distance_threshold = distance_threshold
        self.linkage = linkage
        self.distance = distance
        self.distance_params = distance_params

        super().init()

    def _fit(self, X, y=None):
        """Fit the agglomerative tree to X and cut it to flat clusters.
      """
        self._checkparams(X)
        self.labels = []
        # Implement here
        return self

    def _predict(self, X, y=None) -> np.ndarray:
        """ Implement later"""
        pass
```

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SotonSweetXss**
#### Reference Issues/PRs
Fixes #3433 . See also #1241 

#### What does this implement/fix? Explain your changes.
A `TimeSeriesAgglomerative` has been implemented by using `aeon` to compute a pairwise time series distance matrix (e.g. Euclidean distance, DTW or MSM), and then applying scikit-learn’s `AgglomerativeClustering` algorithm to this pre-computed distance matrix. It supports single-linkage, complete-linkage and average-linkage clustering, but does not support Ward’s linkage clustering, as this method is incompatible with any pre-computed time series distance.

#### Does your contribution introduce a new dependency? If yes, which one?
no, Based entirely on sklearn and Aeon

#### Any other comments?
I will first submit the interface for this level of clustering; if it is approved, I will subsequently add code for reliability verification in `aeon\clustering\test`.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by SotonSweetXss** [comment]
Temporarily changed to Draft PR; experimental comparisons are being added gradually

**Comment by SotonSweetXss** [comment]
The remaining CI failures appear to be an unrelated and unstable test in `test_deep_clusterer_load_model[AEFCNClusterer]`. This has occurred in previous PRs; re-running the test resolved the issue.Please ask a user with write permissions to re-run the failed task.

This test file is mainly used to verify the correctness of `TimeSeriesAgglomerative` when using Euclidean distance.
The tests cover both univariate and multivariate cases. The test data are generated using `make_example_3d_numpy` provided by `aeon`. In each case, the `single`, `complete`, and `average` linkage strategies are tested. For every configuration, the output of `TimeSeriesAgglomerative(distance="euclidean")` is compared with that of scikit-learn's native `AgglomerativeClustering(metric="euclidean")`. The Adjusted Rand Index (ARI)  is used to verify that the two implementations produce equivalent cluster assignments.

In addition, I carried out a simple runtime comparison locally to examine how the runtime changes as the number of samples and the time-series length increase. The experiments also use time-series data generated by `aeon`. They compare scikit-learn clustering directly with Euclidean distance against `TimeSeriesAgglomerative`, which first computes an `aeon` pairwise distance
matrix before performing agglomerative clustering. The figures below show the experimental results.
image
image2

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
thanks for this, sorry for the delay.

1. could you update the __init__ file to include the new clusterer?
2. we need to resolve how to handle predict is empty. I'll get back to you on that
3. testing coverage could be improved
4. _pairwise_distance silently symmetrises the distance matrix
I would prefer a test and raising an exception, since its quite possible we do get an asymmetric distance and we will need to work out what to do rahter than silently "fixing" it
5. File name uses camel case, it should just be _agglomorative, location implies clustering
6. Also add the usual estimator conventions used elsewhere in aeon, such as __all__, __maintainer__, and _get_test_params,

**Comment by SotonSweetXss** [comment]
This update adds `TimeSeriesAgglomerative`, a time series agglomerative
clusterer that computes pairwise distances with aeon and then calls sklearn's
`AgglomerativeClustering` with `metric="precomputed"`.

Main changes:

- added the new clusterer to `aeon.clustering`;
- renamed the implementation file to `_agglomerative.py` to match aeon naming style;
- added aeon estimator conventions, including `__all__`, `__maintainer__`, and
  `_get_test_params`;
- rejected `ward` linkage because it is not compatible with precomputed distances;
- replaced silent distance-matrix symmetrisation with explicit validation and error
  handling;
- added tests for univariate and multivariate time series inputs;
- added Euclidean sanity checks against sklearn;
- added tests for invalid linkage, invalid cluster settings, asymmetric distance
  matrices, and non-finite distance matrices.

I kept the tests focused on the aeon wrapper behavior rather than retesting
sklearn's hierarchical clustering implementation.

**Comment by SotonSweetXss** [comment]
The failing CI checks come from aeon's generic estimator tests calling
`predict`/`predict_proba`. `TimeSeriesAgglomerative` currently raises
`NotImplementedError` in `_predict`, since agglomerative clustering is
transductive and sklearn does not provide a native prediction step for unseen
data.

**Comment by SotonSweetXss** [comment]
I have added `TimeSeriesAgglomerative` to the excluded estimator list in `testing/testing_config` as a short-term fix, since it currently depends on #3613 to pass the generic CI checks, following Tony's suggestion.

**Review by TonyBagnall — APPROVED** [review]
